server/api.go

```diff
@@ -64,30 +64,31 @@ func (p *Plugin) userConnected(w http.ResponseWriter, r *http.Request) {
 }
 
 func (p *Plugin) checkForDMs(userID string) *model.AppError {
-	if p.canSendDiagnostics() {
-		user, err := p.API.GetUser(userID)
-		if err != nil {
-			return err
-		}
+	if !p.canSendDiagnostics() {
+		return nil
+	}
 
-		go func() {
-			// Add a random delay to mitigate against the fact that the user may have multiple sessions hitting this
-			// API at the same time across different servers.
-			p.sleepUpTo(p.userSurveyMaxDelay)
+	now := p.now().UTC()
+	userLockKey := fmt.Sprintf(USER_LOCK_KEY, userID)
 
-			p.connectedLock.Lock()
-			defer p.connectedLock.Unlock()
+	locked, err := p.tryLock(userLockKey, now)
+	if !locked || err != nil {
+		// Either an error occurred or there's already another thread checking for DMs
+		return err
```

Comment:
Nit: Maybe use `errors.Wrap`

Reply:
Forgot about the return value.

Reply:
In my mind you should always use `errors.Wrap`, when yo return an error. The benefit is, that you get detailed trace where this errors was thrown. See https://godoc.org/github.com/pkg/errors#Wrap for an example.

But I forgot about the return value. Hence this is more of a general comment then a request for change.

Reply:
That's definitely a good direction to move in, but I don't think we can do that until we figure out how to stop using `model.AppError` everywhere